File: cpp/286_WallsAndGates.cpp
class Solution {
public:
    vector<vector<int>> directions = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
    void wallsAndGates(vector<vector<int>>& rooms) {
        for (int i = 0; i < rooms.size(); ++i) {
            for (int j = 0; j < rooms[0].size(); ++j) {
                if (rooms[i][j] == 0) {
                    bfs(rooms, i, j);
                }
            }
        }
    }
    
    void bfs(vector<vector<int>>& rooms, int row, int col) {
        queue<vector<int>> q;
        q.push({0, row, col});
        
        while (!q.empty()) {
            vector<int> curr = q.front();
            q.pop();
            
            int i = curr[1], j = curr[2];
            rooms[i][j] = min(rooms[i][j], curr[0]);
            for (vector<int> direction : directions) {
                if (can_go(i + direction[0], j + direction[1], i, j, rooms))
                    q.push({rooms[i][j] + 1, i + direction[0], j + direction[1]});
            }
        }
    }
    
    bool can_go(int i, int j, int i_prior, int j_prior, vector<vector<int>>& rooms) {
        return i >= 0 && i < rooms.size() && j >= 0 && j < rooms[0].size() && rooms[i][j] > 0 && rooms[i][j] > rooms[i_prior][j_prior];
    }
};

File: java/286_WallsAndGates.java
class Solution {
    private static final List<int[]> DIRECTIONS = Arrays.asList(
        new int[] { 1,  0},
        new int[] {-1,  0},
        new int[] { 0,  1},
        new int[] { 0, -1}
    );
    private static final int GATE = 0;

    public void wallsAndGates(int[][] rooms) {
        int m = rooms.length;
        if (m == 0) return;
        int n = rooms[0].length;
        Queue<int[]> q = new LinkedList<>();
        for (int row = 0; row < m; row++) {
            for (int col = 0; col < n; col++) {
                if (rooms[row][col] == GATE) {
                    q.add(new int[] { row, col });
                }
            }
        }
        while (!q.isEmpty()) {
            int[] point = q.poll();
            int i = point[0];
            int j = point[1];
            
            for (int[] direction : DIRECTIONS) {
                int r = i + direction[0];
                int c = j + direction[1];
                if (canGo(r, c, i, j, rooms)) {
                    rooms[r][c] = rooms[i][j] + 1;
                    q.add(new int[] { r, c });
                }
            }
        }
    }
    
    public boolean canGo(int r, int c, int i, int j, int[][] rooms) {
        return r >= 0 && r < rooms.length && c >= 0 && c < rooms[0].length && rooms[r][c] > 0 && rooms[r][c] > rooms[i][j];
    }
}
